### DISCUSSION TITLE
scATAC-scRNA integration

### DISCUSSION THREAD

**Discussion by TauHHCC** [GreenleafLab/ArchR#1959] [Questions / Documentation]
Hi

I am using ArchR to do my scATAC-scRNA integration. 
the cells in my scATAC and scRNA are paired. One cell in scATAC is likely has one relevant cell in scRNA. We would like these cells to integrate together. 
My scATAC is Arch project and my scRNA is Seurat subject. In order to make the integration easier, we make the paired cell same name in both scATAC and scRNA. 
Here is what I did:

I first did the common Unconstrained Integration in Arch.
` scATACintRNA <- addGeneIntegrationMatrix(
  ArchRProj = scATAC, 
  useMatrix = "GeneScoreMatrix",
  matrixName = "GeneIntegrationMatrix",
  reducedDims = "IterativeLSI",
  seRNA = scRNA,
  addToArrow = FALSE,
  groupRNA = 'seurat_clusters',
  nameCell = "scRNA$orig.ident",
  nameGroup = "scRNA@meta.data$seurat_clusters", nameScore = "predictedScore_Un",
  plotUMAP = TRUE
  )`

Then I list the matching cells by their names. 
`matchingCellNames <- intersect(rownames(scATAC@cellColData), colnames(scRNA))`

Then tell the matchingCellNames to groupList
`groupList <- SimpleList(
    Constrained = SimpleList(
        ATAC = matchingCellNames,
        RNA = matchingCellNames
    ),
    Rest = SimpleList(
        ATAC = scATACintRNA$cellNames[!scATACintRNA$cellNames %in% matchingCellNames],
        RNA = colnames(scRNA)[!colnames(scRNA) %in% matchingCellNames]
    )
)`

Finally I did constrained Integration in Arch, constrained by matchingCellNames
`scATACintRNA_0.1<- addGeneIntegrationMatrix(
  ArchRProj = scATACintRNA, 
  useMatrix = "GeneScoreMatrix",
  matrixName = "GeneIntegrationMatrix",
  reducedDims = "IterativeLSI",
  seRNA = scRNA,
  addToArrow = FALSE,
  groupList = groupList,
  groupRNA = 'seurat_clusters',
  nameCell = "scRNA$orig.ident",
  nameGroup = "SCrna@meta.data$seurat_clusters", nameScore = "predictedScore_Un",
  plotUMAP = TRUE
  )`

It went all the process without error, however, then I check the cells that matched. Unfortunately, it was only 23 by Unconstrained Integration and increased to 30 by constrained Integration. Actually, it should be thousands. 

Any place wrong? 

Best

**Comment by uqnsarke** [comment]
I am receiving errors after the integration step,
Error in .safelapply(seq_along(blockList), function(i) { : 
Error Found Iteration 1 : 
	[1] "Error in slot(object = object, name = \"features\")[[layer]] <- features : \n  more elements supplied than there are to replace\n"
	<simpleError in slot(object = object, name = "features")[[layer]] <- features: more elements supplied than there are to replace>
Error Found Iteration 2 : 
	[1] "Error in slot(object = object, name = \"features\")[[layer]] <- features : \n  more elements supplied than there are to replace\n"
	<simpleError in slot(object = object, name = "features")[[layer]] <- features: more elements supplied than there are to replace>
In addition: Warning message:
In mclapply(..., mc.cores = threads, mc.preschedule = preschedule) :
  2 function calls resulted in an error

**Reply by Zepeng-Mu** [reply]
It’s interesting that I had a similar issue recently, although I have never seen this before. Although this is unrelated to the main topic of this discussion thread. https://github.com/GreenleafLab/ArchR/issues/1999#issuecomment-1761987935
But basically after a lot of digging around, I think what might be happening is that in some cases the colnames of the peak matrix is an array type instead of character, and this leads to the above error in Seurat. I have run this function multiple times, so I don’t know why it’s happening now. But I basically added a line of code that converts the colnames to character before createSeuratObject function and everything has been fine so far.

hope this is helpful

**Reply by uqnsarke** [reply]
Thanks for your prompt response. Would you mind to reshare your code?
